### ISSUE TITLE
some cluster were not migrated on cloud

### ISSUE THREAD

**Issue by p0lyn0mial** [kubermatic/kubermatic#2022]
**Description**

the migration tool reports that some clusters were actually not migrated on `cloud` environment. 

```
I0918 15:30:29.961116    4161 main.go:658] STEP 5: migrating cluster resources
W0918 15:30:29.961126    4161 main.go:665] skipping cluster resource = nz8mqphvhx (physical location seed/us-central1-c-1), there is no project for a user with ID = ChUxMTY3Mjc2OTE3OTEzOTA0ODIxMjgSBmdvb2dsZQ (the user doesn't exist ?)
W0918 15:30:29.961136    4161 main.go:665] skipping cluster resource = w898qbsrrf (physical location seed/us-central1-c-1), there is no project for a user with ID = ChUxMTY3Mjc2OTE3OTEzOTA0ODIxMjgSBmdvb2dsZQ (the user doesn't exist ?)
W0918 15:30:29.961144    4161 main.go:665] skipping cluster resource = vqpdwq9c7c (physical location seed/us-central1-c-1), there is no project for a user with ID = ChUxMDI3MzczNjc5NzIyODM4Mjc3NjISBmdvb2dsZQ (the user doesn't exist ?)
W0918 15:30:29.961154    4161 main.go:665] skipping cluster resource = 5kj6kl9lxz (physical location seed/asia-east1-a-1), there is no project for a user with ID = ChUxMDM1NjQ5NTExNzUxODQ3MTY2MzQSBmdvb2dsZQ (the user doesn't exist ?)
I0918 15:30:29.961164    4161 main.go:677] the cluster resource (Name = mbm4zbhnzf) for userID = ChUxMDEwMDAxMDczNjc4MDA4MzA2NjESBmdvb2dsZQ was NOT migrated because dry-run option was requested, physical location = seed/europe-west3-c-1
W0918 15:30:29.961174    4161 main.go:665] skipping cluster resource = hnsvkngxsr (physical location seed/us-central1-c-1), there is no project for a user with ID = ChUxMTU2NDgyNTMxOTg1NDY0OTMyNjMSBmdvb2dsZQ (the user doesn't exist ?)
```

I did a small research and there is no user resource for `vqpdwq9c7c` and `hnsvkngxsr` cluster. In the old model there was a special label attached to clusters that connected a cluster with a user, see:
```
k get cluster vqpdwq9c7c -oyaml
kubectl get users -o json | jq -r '[.items[] | {email:.spec.email}]' | grep "15ngburton@gmail.com"
```

moreover there are users for clusters `w898qbsrrf ` and `nz8mqphvhx` cluster but with a different ID. For example `k get cluster w898qbsrrf -oyaml` returns:

```
...
  labels:
    user: ChUxMTY3Mjc2OTE3OTEzOTA0ODIxMjgSBmdvb2dsZQ
  userEmail: chrisgaun@gmail.com
  userName: Chris Gaun
```

issuing  `kubectl get users -o json | jq -r '[.items[] | {id:.spec.id, name:.metadata.name, email:.spec.email}]'` yields:
```
  {
    "id": "CggxMDE3MDA2ORIGZ2l0aHVi",
    "name": "user-zl5b5",
    "email": "chrisgaun@gmail.com"
  }
```

**Action**
- remove cluster without the owner 
- manually update labels on `w898qbsrrf ` and `nz8mqphvhx` clusters so that they are migrated ?

**Comment by p0lyn0mial** [comment]
Non existing cluster `vqpdwq9c7c`:
 
There is no user for this cluster in our system. However we may create a user resource when the user actually logs in. If the `ID` is the same then the cluster is going to be migrated the next time the migration tool is run.

**Comment by p0lyn0mial** [comment]
Non existing cluster `hnsvkngxsr` and `5kj6kl9lxz`:

as above.

**Comment by p0lyn0mial** [comment]
Cluster `w898qbsrrf` and user `user-zl5b5`, `ID` mismatch:
 
The Cluster has`ChUxMTY3Mjc2OTE3OTEzOTA0ODIxMjgSBmdvb2dsZQ`, the user has `CggxMDE3MDA2ORIGZ2l0aHVi`

The cluster was created on `2018-06-05T20:56:01Z`, the user was created on `2018-07-26T03:45:21Z` 

In the old model we read the ID (`sub`) from the token and add it to the newly created clusters.
The issue with that is that the ID('sub') is not unique and will change depending on the issuer.
For example it will be different when you authenticate with Google or GitHub even though they map to the same email address. See also: https://tools.ietf.org/html/rfc7519

Some time ago I added  code that actually updates the ID both in the context and in the database https://github.com/kubermatic/kubermatic/commit/dee92259207d15d479caaab5dd05fcfc323fd0a2#diff-59a8d107efa4838cdd8127f7a18407b0. This only slightly improves the situation as there will be match between the ID in the clusters resources and what we store in the database. But it will flip every time a decides to use a different issuer.

It doesn't affect the new model, except that occasionally there will be `409` as subsequent request are trying to update `user` resource. 

To sum up, I think that in the old model it was broken from the very beginning and the fact that no-ones complains makes me feel that the users don't filip between the issuers. It would leave the code as it is now and recommend not to change the issuers. Once we will roll out the new version I would remove the code that updates the user resource.

/cc @mrIncompetent

decided to manually update the IDs so that they match.

**Comment by mrIncompetent** [comment]
Before we delete clusters without owners, we must ensure that those clusters are "dead"
With dead I mean: No nodes & nothing custom deployed in the clusters

**Comment by p0lyn0mial** [comment]
@mrIncompetent okay, let me check if we can remove them.

**Comment by p0lyn0mial** [comment]
cluster `hnsvkngxsr`, no nodes but some pods deployed

```
k get node 
No resources found.
```

```
k get po --all-namespaces=true                                  
NAMESPACE     NAME                                     READY     STATUS    RESTARTS   AGE
deis          deis-builder-55bbd946bd-q7t6g            0/1       Pending   0          138d
deis          deis-controller-59c948dbd4-xdjxc         0/1       Pending   0          138d
deis          deis-database-5875dcff85-m5nql           0/1       Pending   0          138d
deis          deis-logger-7445666f95-zgckp             0/1       Pending   0          138d
deis          deis-logger-redis-7fcc95cc5d-c8v64       0/1       Pending   0          138d
deis          deis-minio-64454c6bf6-4zg2d              0/1       Pending   0          138d
deis          deis-monitor-grafana-77d54db485-qjzt9    0/1       Pending   0          138d
deis          deis-monitor-influxdb-5fb56d489c-z7nvp   0/1       Pending   0          138d
deis          deis-nsqd-649fbc4947-gx8mm               0/1       Pending   0          138d
deis          deis-registry-554c5569db-4lvxg           0/1       Pending   0          138d
deis          deis-router-98ccf5c64-qdljj              0/1       Pending   0          138d
deis          deis-workflow-manager-76566c6cc-2625k    0/1       Pending   0          138d
kube-system   gitkube-controller-664fb484b5-w2jxd      0/1       Pending   0          138d
kube-system   gitkubed-84456dcbb9-gh94k                0/1       Pending   0          138d
kube-system   ill-duck-traefik-5cdd44f798-c79p4        0/1       Pending   0          138d
kube-system   kube-dns-774d5484cc-5cjl5                0/3       Pending   0          138d
kube-system   kube-dns-autoscaler-69c5cbdcdd-dcgtg     0/1       Pending   0          138d
kube-system   kubernetes-dashboard-74f855c8c6-682sk    0/1       Pending   0          57d
kube-system   kubernetes-dashboard-78f8d5f498-rrcnr    0/1       Pending   0          41d
kube-system   metrics-server-55b6f4b9cd-7pg2m          0/1       Pending   0          57d
kube-system   openvpn-client-5f5954f465-nzxwf          0/2       Pending   0          4d
kube-system   tiller-deploy-7bf964fff8-fxm4c           0/1       Pending   0          138d

```

```
k describe po -n=deis deis-builder-55bbd946bd-q7t6g                                                                                                                                                                                                                                                                    
Name:           deis-builder-55bbd946bd-q7t6g
Namespace:      deis
Node:           <none>
Labels:         app=deis-builder
                pod-template-hash=1166850268
Annotations:    <none>
Status:         Pending
IP:             
Controlled By:  ReplicaSet/deis-builder-55bbd946bd
Containers:
  deis-builder:
    Image:      quay.io/deis/builder:v2.12.0
    Ports:      2223/TCP, 8092/TCP
    Liveness:   http-get http://:8092/healthz delay=30s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:8092/readiness delay=30s timeout=1s period=10s #success=1 #failure=3
    Environment:
      DEIS_REGISTRY_SERVICE_HOST:        127.0.0.1
      DEIS_REGISTRY_SERVICE_PORT:        5555
      HEALTH_SERVER_PORT:                8092
      EXTERNAL_PORT:                     2223
      BUILDER_STORAGE:                   minio
      DEIS_REGISTRY_LOCATION:            on-cluster
      DEIS_REGISTRY_SECRET_PREFIX:       private-registry
      GIT_LOCK_TIMEOUT:                  10
      SLUGBUILDER_IMAGE_NAME:            <set to the key 'image' of config map 'slugbuilder-config'>         Optional: false
      SLUG_BUILDER_IMAGE_PULL_POLICY:    <set to the key 'pullpolicy' of config map 'slugbuilder-config'>    Optional: false
      DOCKERBUILDER_IMAGE_NAME:          <set to the key 'image' of config map 'dockerbuilder-config'>       Optional: false
      DOCKER_BUILDER_IMAGE_PULL_POLICY:  <set to the key 'pullpolicy' of config map 'dockerbuilder-config'>  Optional: false
      DOCKERIMAGE:                       1
      DEIS_DEBUG:                        false
      POD_NAMESPACE:                     deis (v1:metadata.namespace)
      DEIS_BUILDER_KEY:                  <set to the key 'builder-key' in secret 'builder-key-auth'>  Optional: false
    Mounts:
      /var/run/secrets/api/auth from builder-key-auth (ro)
      /var/run/secrets/deis/builder/ssh from builder-ssh-private-keys (ro)
      /var/run/secrets/deis/objectstore/creds from objectstore-creds (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from deis-builder-token-tq9sb (ro)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  builder-key-auth:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  builder-key-auth
    Optional:    false
  builder-ssh-private-keys:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  builder-ssh-private-keys
    Optional:    false
  objectstore-creds:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  objectstorage-keyfile
    Optional:    false
  deis-builder-token-tq9sb:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  deis-builder-token-tq9sb
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason            Age                         From               Message
  ----     ------            ----                        ----               -------
  Warning  FailedScheduling  <invalid> (x24331 over 4d)  default-scheduler  no nodes available to schedule pods
```

**Comment by p0lyn0mial** [comment]
cluster `vqpdwq9c7c` no nodes a no custom workloads
```
k get po --all-namespaces=true                                
NAMESPACE     NAME                                    READY     STATUS    RESTARTS   AGE
kube-system   kube-dns-774d5484cc-72cfk               0/3       Pending   0          137d
kube-system   kube-dns-autoscaler-69c5cbdcdd-r57br    0/1       Pending   0          137d
kube-system   kubernetes-dashboard-74f855c8c6-sdstn   0/1       Pending   0          58d
kube-system   kubernetes-dashboard-78f8d5f498-5rgkc   0/1       Pending   0          41d
kube-system   metrics-server-55b6f4b9cd-xjtwh         0/1       Pending   0          58d
kube-system   openvpn-client-5f5954f465-9rz54         0/2       Pending   0          5d
```

```
k get node                    
No resources found.
```

```
k get machines -n=kube-system
No resources found.
```

```
k config view                
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: REDACTED
    server: https://vqpdwq9c7c.us-central1-c-1.cloud.kubermatic.io:31263
  name: vqpdwq9c7c
```

**Comment by p0lyn0mial** [comment]
cluster `5kj6kl9lxz` no nodes but some pods deployed

```
k get po --all-namespaces=true 
NAMESPACE     NAME                                    READY     STATUS    RESTARTS   AGE
default       mongo-express-667d756d68-c7ss4          0/1       Pending   0          142d
default       pbook-back-84d5f4b876-2hbc5             0/1       Pending   0          142d
default       pbook-db-67b4964778-26b2h               0/1       Pending   0          142d
default       pbook-front-797d6965dd-7f28n            0/1       Pending   0          142d
default       pbook-proxy-69579d5b-ljzl4              0/1       Pending   0          142d
kube-system   kube-dns-774d5484cc-bslh7               0/3       Pending   0          142d
kube-system   kube-dns-774d5484cc-qms6w               0/3       Pending   0          142d
kube-system   kube-dns-autoscaler-69c5cbdcdd-rrd7z    0/1       Pending   0          142d
kube-system   kubernetes-dashboard-74f855c8c6-f2xlx   0/1       Pending   0          58d
kube-system   kubernetes-dashboard-78f8d5f498-l8lzq   0/1       Pending   0          41d
kube-system   metrics-server-55b6f4b9cd-vk7w7         0/1       Pending   0          58d
kube-system   openvpn-client-5f5954f465-t9wsj         0/2       Pending   0          5d
```

```
k get node                    
No resources found.

```

```
k get machines -n=kube-system
No resources found.
```

```
k config view                      
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: REDACTED
    server: https://5kj6kl9lxz.asia-east1-a-1.cloud.kubermatic.io:31518
  name: 5kj6kl9lxz
```

**Comment by p0lyn0mial** [comment]
@mrIncompetent let me know when you would like to write an email to the `vqpdwq9c7c ` and `5kj6kl9lxz ` cluster owners

**Comment by mrIncompetent** [comment]
I deleted those cluster, also i deleted all ssh keys, that caused error messages in the rbac controller
